cns/service.go

```diff
@@ -48,64 +47,93 @@ func NewService(name, version, channelMode string, store store.KeyValueStore) (*
 	}, nil
 }
 
-// GetAPIServerURL returns the API server URL.
-func (service *Service) getAPIServerURL() string {
-	urls, _ := service.GetOption(acn.OptCnsURL).(string)
-	if urls == "" {
-		urls = defaultAPIServerURL
-	}
-
-	return urls
-}
-
-// Initialize initializes the service and starts the listener.
-func (service *Service) Initialize(config *common.ServiceConfig) error {
-	log.Debugf("[Azure CNS] Going to initialize a service with config: %+v", config)
-
-	// Initialize the base service.
-	if err := service.Service.Initialize(config); err != nil {
-		return errors.Wrap(err, "failed to initialize")
-	}
+func (service *Service) AddListener(config *common.ServiceConfig) error {
```

Comment:
Much of this is massaging the config into derived configuration, right? It would probably make more sense to define methods on the config to handle that.